File: core/src/passes/branch_fold.cpp
// core/src/passes/branch_fold.cpp — statically-resolve trivial conds.
//
// When `CmpFlags const_a, const_b` is followed immediately by
// `CondJumpRel cc, taken, fallthrough`, the branch's direction is
// known at compile time. This pass rewrites the CondJumpRel to a
// plain JumpRel at whichever side the condition picks.
//
// Works on a single-pass scan. The CmpFlags isn't removed — dropping
// it would change observable flag state that a later instruction
// might read. DCE handles the actual removal once the flag state is
// provably unused.

#include "prisma/passes.hpp"

#include <cstdint>
#include <optional>
#include <unordered_map>
#include <variant>

namespace prisma::passes {

namespace {

struct KnownConst {
    std::uint64_t value;
    ir::OpSize    size;
};

// Evaluate `a cc b` on two 64-bit values with size-specific masking.
// Returns true/false when the condition can be decided. Returns nullopt
// for flag-direct cases whose carry/overflow/sign semantics are not
// modelled by this MVP pass.
std::optional<bool> eval_cond(ir::CondCode cc,
                              std::uint64_t a,
                              std::uint64_t b,
                              ir::OpSize size) noexcept {
    // Mask to operand size — the comparison sees truncated values.
    a = ir::mask_to_size(a, size);
    b = ir::mask_to_size(b, size);

    // Signed interpretation for the *-signed comparisons. Extract the
    // value's bit width from the size enum so we sign-extend correctly.
    const std::uint32_t bits = ir::bit_width(size);
    auto sext = [&](std::uint64_t v) -> std::int64_t {
        if (bits == 64) return static_cast<std::int64_t>(v);
        const std::uint64_t sign_bit = 1ULL << (bits - 1);
        if (v & sign_bit) {
            const std::uint64_t mask = ~((1ULL << bits) - 1);
            return static_cast<std::int64_t>(v | mask);
        }
        return static_cast<std::int64_t>(v);
    };

    const std::int64_t sa = sext(a);
    const std::int64_t sb = sext(b);

    switch (cc) {
        case ir::CondCode::Eq:  return a == b;
        case ir::CondCode::Ne:  return a != b;
        case ir::CondCode::Ult: return a <  b;
        case ir::CondCode::Ule: return a <= b;
        case ir::CondCode::Ugt: return a >  b;
        case ir::CondCode::Uge: return a >= b;
        case ir::CondCode::Slt: return sa <  sb;
        case ir::CondCode::Sle: return sa <= sb;
        case ir::CondCode::Sgt: return sa >  sb;
        case ir::CondCode::Sge: return sa >= sb;
        // Flag-direct cases (Cc / Nc / Ov / NoOv / Mi / Pl) can't be
        // decided without modelling the full NZCV update — leave them
        // unfolded for now.
        case ir::CondCode::Cc:
        case ir::CondCode::Nc:
        case ir::CondCode::Ov:
        case ir::CondCode::NoOv:
        case ir::CondCode::Mi:
        case ir::CondCode::Pl:
            return std::nullopt;
    }
    return std::nullopt;
}

}  // namespace

std::vector<ir::Stmt>
branch_fold(const std::vector<ir::Stmt>& stmts) {
    std::unordered_map<ir::Ref, KnownConst> consts;

    // Track the most recent CmpFlags so the next CondJumpRel can
    // consult it. Cleared on any flag-clobbering op.
    struct LastCmp {
        ir::Ref  lhs;
        ir::Ref  rhs;
        ir::OpSize size;
    };
    std::optional<LastCmp> last_cmp;

    std::vector<ir::Stmt> out;
    out.reserve(stmts.size());

    for (const auto& s : stmts) {
        if (std::holds_alternative<ir::Constant>(s.op) && s.result) {
            const auto& c = std::get<ir::Constant>(s.op);
            consts[*s.result] = KnownConst{
                ir::mask_to_size(c.value, c.size), c.size};
        }
        else if (std::holds_alternative<ir::CmpFlags>(s.op)) {
            const auto& cf = std::get<ir::CmpFlags>(s.op);
            last_cmp = LastCmp{cf.lhs, cf.rhs, cf.size};
        }
        else if (std::holds_alternative<ir::CondJumpRel>(s.op) && last_cmp) {
            const auto& cj = std::get<ir::CondJumpRel>(s.op);
            const auto it_l = consts.find(last_cmp->lhs);
            const auto it_r = consts.find(last_cmp->rhs);
            if (it_l != consts.end() && it_r != consts.end()) {
                const auto taken = eval_cond(
                    cj.cc, it_l->second.value, it_r->second.value,
                    last_cmp->size);
                if (taken) {
                    ir::Stmt folded{std::nullopt,
                        ir::Op{ir::JumpRel{
                            *taken ? cj.target_guest_pc : cj.fallthrough_guest_pc}}};
                    out.push_back(std::move(folded));
                    last_cmp.reset();
                    continue;
                }
            }
            last_cmp.reset();
        }
        // Any ALU op that writes could invalidate NZCV; conservatively
        // flush last_cmp when any BinOp or StoreReg touches flags-
        // adjacent state. We're over-conservative here because the IR
        // doesn't yet model flag-setting variants.
        else if (std::holds_alternative<ir::BinOp>(s.op)) {
            // BinOp does NOT clobber flags in our IR today — only
            // CmpFlags does — so leave last_cmp alone.
        }

        out.push_back(s);
    }

    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/branch_fold.rs
//! Branch folding pass.
//!
//! Mirrors C++ `branch_fold`. When a `CmpFlags const_a, const_b` is followed
//! (with no intervening flag clobber) by a `CondJumpRel`, the direction is
//! known at compile time and the `CondJumpRel` is rewritten to a plain
//! `JumpRel`. The `CmpFlags` is left in place; DCE removes it once the flag
//! state is provably unused. Flag-direct conditions (carry/overflow/sign) are
//! not folded since the IR does not model the full NZCV update here.

use std::collections::HashMap;

use prisma_ir::{BasicBlock, CondCode, Function, JumpRel, Op, OpSize, Ref, Stmt};

use crate::Pass;

/// Evaluate `a cc b` with size-specific masking; `None` for flag-direct ccs.
fn eval_cond(cc: CondCode, a: u64, b: u64, size: OpSize) -> Option<bool> {
    let a = a & size.mask();
    let b = b & size.mask();
    let bits = size.bit_width();
    let sext = |v: u64| -> i64 {
        if bits == 64 {
            v as i64
        } else {
            let sign_bit = 1u64 << (bits - 1);
            if v & sign_bit != 0 {
                (v | !((1u64 << bits) - 1)) as i64
            } else {
                v as i64
            }
        }
    };
    let (sa, sb) = (sext(a), sext(b));
    match cc {
        CondCode::Eq => Some(a == b),
        CondCode::Ne => Some(a != b),
        CondCode::Ult => Some(a < b),
        CondCode::Ule => Some(a <= b),
        CondCode::Ugt => Some(a > b),
        CondCode::Uge => Some(a >= b),
        CondCode::Slt => Some(sa < sb),
        CondCode::Sle => Some(sa <= sb),
        CondCode::Sgt => Some(sa > sb),
        CondCode::Sge => Some(sa >= sb),
        CondCode::Cc
        | CondCode::Nc
        | CondCode::Ov
        | CondCode::NoOv
        | CondCode::Mi
        | CondCode::Pl => None,
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct BranchFold;

impl BranchFold {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for BranchFold {
    fn name(&self) -> &'static str {
        "branch_fold"
    }

    fn run(&self, func: Function) -> Function {
        branch_fold(func)
    }
}

#[derive(Clone, Copy)]
struct LastCmp {
    lhs: Ref,
    rhs: Ref,
    size: OpSize,
}

/// Per-block branch folding.
pub fn branch_fold(func: Function) -> Function {
    let blocks = func
        .blocks
        .into_iter()
        .map(|block| {
            let mut consts: HashMap<Ref, u64> = HashMap::new();
            let mut last_cmp: Option<LastCmp> = None;
            let mut out = Vec::with_capacity(block.stmts.len());

            for stmt in block.stmts {
                match &stmt.op {
                    Op::Constant(c) => {
                        if let Some(r) = stmt.result {
                            consts.insert(r, c.value & c.size.mask());
                        }
                    }
                    Op::CmpFlags(cf) => {
                        last_cmp = Some(LastCmp {
                            lhs: cf.lhs,
                            rhs: cf.rhs,
                            size: cf.size,
                        });
                    }
                    Op::AluFlags(_) | Op::WriteFlagsCountZero(_) => {
                        last_cmp = None;
                    }
                    Op::CondJumpRel(cj) => {
                        if let Some(lc) = last_cmp {
                            if let (Some(&a), Some(&b)) = (consts.get(&lc.lhs), consts.get(&lc.rhs))
                            {
                                let choose_taken = eval_cond(cj.cc, a, b, lc.size).unwrap_or(false);
                                let target = if choose_taken {
                                    cj.target_guest_pc
                                } else {
                                    cj.fallthrough_guest_pc
                                };
                                out.push(Stmt::new(
                                    None,
                                    Op::JumpRel(JumpRel {
                                        target_guest_pc: target,
                                    }),
                                ));
                                last_cmp = None;
                                continue;
                            }
                        }
                        last_cmp = None;
                    }
                    _ => {}
                }
                out.push(stmt);
            }

            BasicBlock {
                id: block.id,
                stmts: out,
            }
        })
        .collect();

    Function {
        blocks,
        entry: func.entry,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{CmpFlags, CondJumpRel, Constant};

    fn cst(result: u32, value: u64) -> Stmt {
        Stmt::new(
            Some(result),
            Op::Constant(Constant {
                value,
                size: OpSize::I64,
            }),
        )
    }

    fn build(cc: CondCode, a: u64, b: u64) -> Function {
        Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    cst(0, a),
                    cst(1, b),
                    Stmt::new(
                        None,
                        Op::CmpFlags(CmpFlags {
                            lhs: 0,
                            rhs: 1,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        None,
                        Op::CondJumpRel(CondJumpRel {
                            cc,
                            target_guest_pc: 0x1000,
                            fallthrough_guest_pc: 0x2000,
                        }),
                    ),
                ],
            }],
        }
    }

    fn last_target(func: &Function) -> Option<u64> {
        match &func.blocks[0].stmts.last().unwrap().op {
            Op::JumpRel(j) => Some(j.target_guest_pc),
            _ => None,
        }
    }

    #[test]
    fn folds_taken_branch() {
        // 5 == 5 -> taken -> target 0x1000
        let out = branch_fold(build(CondCode::Eq, 5, 5));
        assert_eq!(last_target(&out), Some(0x1000));
    }

    #[test]
    fn folds_not_taken_branch() {
        // 5 == 6 -> false -> fallthrough 0x2000
        let out = branch_fold(build(CondCode::Eq, 5, 6));
        assert_eq!(last_target(&out), Some(0x2000));
    }

    #[test]
    fn signed_less_than_uses_signed_interpretation() {
        // -1 (0xFFFF..) Slt 1 -> true -> target
        let out = branch_fold(build(CondCode::Slt, u64::MAX, 1));
        assert_eq!(last_target(&out), Some(0x1000));
        // unsigned would be false: confirm Ult picks fallthrough.
        let out2 = branch_fold(build(CondCode::Ult, u64::MAX, 1));
        assert_eq!(last_target(&out2), Some(0x2000));
    }

    #[test]
    fn flag_direct_cc_folds_to_fallthrough() {
        // eval_cond returns None for flag-direct ccs; with both operands known
        // the C++ pass folds via unwrap_or(false) -> fallthrough. Match it.
        let out = branch_fold(build(CondCode::Cc, 5, 5));
        assert_eq!(last_target(&out), Some(0x2000));
    }

    #[test]
    fn non_constant_operands_leave_branch() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(
                        None,
                        Op::CmpFlags(CmpFlags {
                            lhs: 7,
                            rhs: 8,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        None,
                        Op::CondJumpRel(CondJumpRel {
                            cc: CondCode::Eq,
                            target_guest_pc: 0x1000,
                            fallthrough_guest_pc: 0x2000,
                        }),
                    ),
                ],
            }],
        };
        let out = branch_fold(func.clone());
        assert_eq!(out, func);
    }
}
